### DISCUSSION TITLE
Trying to split metadata into artist and title

### DISCUSSION THREAD

**Discussion by gavin2812** [savonet/liquidsoap#3392] [User Support]
Hello,

I'm trying to simply split the m["title" into artist and title using the script below, but all I'm getting is the following error (i'm using 2.1.0):
```
At /home/test/metadata.liq, line 12, char 25-37:
	  let (artist, title) = string.split.first(separator=" - ", metatitle)
Error 5: this value has no method `first`
```

Code:
```
if string.contains(substring=" - ",m["title"]) then
	  let (artist, title) = string.split.first(separator=" - ", m["title"])
	  print("artist: #{artist}")
	  print("title: #{title}")
	end
```

I've tried with variations such as:

```
def print_metadata(m) =
  metatitle = m["title"]
	if string.contains(substring=" - ",metatitle) then
	  let (artist, title) = string.split.first(separator=" - ", metatitle)
	  print("artist: #{artist}")
	  print("title: #{title}")
	end
end
radio.on_metadata(print_metadata)
```
I've also tried this:
```
if string.contains(substring=" - ", m["title"]) then
      let (a, t) = string.split.first(separator=" - ", m["title"])
      nowplaying.set(shorten(t) ^ "\n" ^ shorten(a))
    end
```
Which also gives an error:
```
At /home/test/metadata.liq, line 20, char 19-31:
      let (a, t) = string.split.first(separator=" - ", m["title"])
Error 5: this value has no method `first`
```
Can anyone help with what I'm doing wrong?

Thanks.

**Comment by gAlleb** [comment]
@gavin2812 
Hi, and what exactly are you trying to achieve? Metadata is already divided into artist and title. It is Icecast which treats 'artist' and 'title' as one line. You can put metadata into file for example and it would be splitted in 2 lines:

``` 
def write_meta(m) =
ignore(file.write(data="#{m[\"artist\"]}\n#{m[\"title\"]}", append=false, "/path/to/metadata.php"))
end
radio.on_metadata(write_meta)
```

it will result in this:
Artist
Title

Or, you can fetch original metadata 'Artist - title'

```
def write_meta(m) =
ignore(file.write(data="#{m[\"artist\"]} - #{m[\"title\"]}", append=false, "/path/to/metadata.php"))
end
```

And do with it whatever you like: split it using php e.g:

```
$url = "metadata.php";
		$data = file_get_contents($url);
			$currently_playing = $data;
			$currently_playing = explode(" - ", $currently_playing, 2);
			$array['Title'] = $currently_playing[1];
			$array['Artist'] = explode(";",$currently_playing[0])[0];
			echo json_encode($array);
```

or split it with js:

```
var xhttp = new XMLHttpRequest();
    xhttp.onreadystatechange = function () {
this.responseText.search("-")!=-1;
            let artist = this.responseText.substring(0,this.responseText.search("-")).trim();
            let title = this.responseText.substring(this.responseText.search("-")+1).trim();
}
 xhttp.open('GET', 'path/to/metadata.php', true);
 xhttp.send();
```

**Comment by gavin2812** [comment]
@gAlleb Thanks for your reply, much appreciated

The issue is that I only have both the artist and title in m["title"] and I need to separate them.

I'm aware I could manipulate the data outside of the liquidsoap script (using PHP for example) but I want to do it all via the liquidsoap script without any other code.

Ideally I just need to be able to split the contents of m["title"] using the - (hyphen) delimiter and obtain two separate variables/strings which contain only the artist and (separately) the song title.

**Reply by gAlleb** [reply]
I see. In order for you to use `string.split.first` you need to upgrade to latest version of Liquidsoap because `string.split.first` has been added not so long ago and 2.1.0 doesn't have this option.

See https://github.com/savonet/liquidsoap/discussions/3145 [https://github.com/savonet/liquidsoap/discussions/3145](url)

**Reply by gAlleb** [reply]
If you wanna try, then fully working code below.  
```
currently_playing = ref("")
currently_playing2 = ref("")

def update_metadata(m)   
    if string.length(m["artist"]) <= 0 then
      if string.contains(substring=" - ", m["title"]) then
          let (a, t) = string.split.first(separator=" - ", m["title"])
          currently_playing.set(t)
	  currently_playing2.set(a)
      end
     end
end
radio.on_metadata(update_metadata)
```
Example from my script. I assume you want to decode Icecast title tag into two parts. Below is my example: Icecast stream relay (input.http) into HLS live.m3u8. Icecast transmits only 'title', but now it is seperated into `artist` (currently_playing2.set(a)) and `title` (currently_playing.set(t)). Liq 2.2.0

<img width="500" alt="Снимок экрана 2023-09-09 в 20 28 40" src="https://github.com/savonet/liquidsoap/assets/144447266/f9f3fb6d-034e-4077-9a68-f1a716b32bfe">

**Comment by vitoyucepi** [comment]
Hi @gavin2812,
You should definitely avoid combining artist and title in the first place because there are some cases where the resulting string contains more than one dash sequence. Imagine the following string `The Folk - Poppers feat. Lizzy Blue - Sky - Heartache - Dancing`. Perhaps it's not as terrible as I thought since I purposely included additional spaces before and after the dashes. After all, the real name could be `The Folk-Poppers feat. Lizzy Blue-Sky - Heartache-Dancing`.

Anyway here's the required function.
I've added the `string.split.first` function polyfill that will provide a similar experience, but sometimes the result could be different.
```ruby
%ifndef string.split.first
def string.split.first(~separator, s)
    split_parts =  string.split(separator=separator, s)
    if list.length(split_parts) == 0 then
        error.raise(error.not_found, "String does not contain the separator.")
    end
    first = list.hd(split_parts)
    rest = string.concat(list.tl(split_parts), separator=separator)
    (first, rest)
end
%endif

def split_dash(_string)
    try
        string.split.first(_string, separator=" - ")
    catch _ do
        (_string, "")
    end
end
```

Here are the tests.
```ruby
test = [
    "The Folk - Poppers feat. Lizzy Blue - Sky - Heartache - Dancing",
    "The Folk-Poppers feat. Lizzy Blue-Sky - Heartache-Dancing",
    "The Folk-Poppers - Heartache-Dancing feat. Lizzy Blue-Sky",
    "The FolkPoppers feat. Lizzy BlueSky - HeartacheDancing",
    "The FolkPoppers",
    "HeartacheDancing",
    " - HeartacheDancing",
    ""
]
expected = [
    ("The Folk", "Poppers feat. Lizzy Blue - Sky - Heartache - Dancing"),
    ("The Folk-Poppers feat. Lizzy Blue-Sky", "Heartache-Dancing"),
    ("The Folk-Poppers", "Heartache-Dancing feat. Lizzy Blue-Sky"),
    ("The FolkPoppers feat. Lizzy BlueSky", "HeartacheDancing"),
    ("The FolkPoppers", ""),
    ("HeartacheDancing", ""),
    ("", "HeartacheDancing"),
    ("", "")
]

def check(x, y)
    if x != y then
        print("Failure: got #{x} instead of #{y}")
    end
end
result = list.map(split_dash, test)
check(result, expected)
```
